Docs/isd 5129: Add Tutorial Documentation (#52)
canonical/mattermost-k8s-operator: docs/tutorial.rst

complete this step for you, and you can verify by checking for ``msg="Bootstrapped Juju" provider=microk8s``
in the terminal output and by running ``juju controllers``.

If Concierge did not perform the bootstrap, run:

.. code-block:: bash

   juju bootstrap microk8s tutorial-controller

What you'll do
--------------

1. Deploy the Mattermost charm
2. Integrate with the PostgreSQL charm
3. Inspect the Kubernetes resources created
4. Access the Mattermost app and bootstrap
5. Clean up the environment

Add a Juju model for the tutorial
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To easily clean up the resources and separate your workload from the contents of this tutorial, set up a new Juju model named ``mattermost-tutorial``:

.. code-block:: bash

   juju add-model mattermost-tutorial

Deploy the charms
~~~~~~~~~~~~~~~~~

Mattermost requires connections to PostgreSQL. For more information, see the `Charm Integrations <https://charmhub.io/mattermost-k8s/docs/reference-integrations>`_.

Deploy the charms:

.. code-block:: bash

   juju deploy mattermost-k8s --channel latest/edge
   juju deploy postgresql-k8s --channel 14/stable --trust

> The Mattermost database driver requires a secure SSL/TLS connection by default. For this tutorial, we will use the ``self-signed-certificates`` charm to provision a local Certificate Authority(CA) for PostgreSQL:

Reviewer: ```suggestion
The Mattermost database driver requires a secure SSL/TLS connection by default. For this tutorial, we will use the ``self-signed-certificates`` charm to provision a local Certificate Authority (CA) for PostgreSQL:
```
the nittest of picks

   postgresql-k8s:upgrade                 postgresql-k8s:upgrade         upgrade            peer
   self-signed-certificates:certificates  postgresql-k8s:certificates    tls-certificates   regular

The deployment finishes when all the charms show ``Active`` states.

Run ``microk8s kubectl get pods -n mattermost-tutorial`` to see the pods that are being created by the charms:

.. code-block:: text

   NAME                            READY   STATUS    RESTARTS   AGE
   mattermost-k8s-0                2/2     Running   0          10m
   modeloperator-64cb49db9-8cvpv   1/1     Running   0          15m
   postgresql-k8s-0                2/2     Running   0          13m
   self-signed-certificates-0      1/1     Running   0          13m

.. note::

   If you are using ``multipass``, or installed ``microk8s`` from scratch, you might get an "insufficient permissions" error. In that case, run the commands presented at the error message in your terminal, after which you'll be able to run ``microk8s`` commands.

   .. code-block:: bash

      sudo usermod -a -G snap_microk8s ubuntu
      sudo chown -R ubuntu ~/.kube
      newgrp snap_microk8s

Access ``Mattermost`` for the first time

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: Nit: I think we don't need to code-format Mattermost here
```suggestion
Access Mattermost for the first time
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

By default, ``mattermost`` is exposed on the port 8080. To find the internal IP address assigned to ``mattermost``, check the application address in the ``juju status`` output. For our example, this is ``10.152.183.239``. Open a web browser, navigate to ``10.152.183.239:8080`` and follow the steps to set up your Mattermost server.

.. note::

>    If you are using ``multipass``, you need to forward the port to access the application from a web browser:

Reviewer: Nitpick: I don't think we need to code-format Multipass here:
```suggestion
   If you are using Multipass, you need to forward the port to access the application from a web browser:
```

   .. code-block:: bash

      microk8s kubectl port-forward --address 0.0.0.0 service/mattermost-k8s 8080:8080 -n mattermost-tutorial

   Then, in a separate terminal in your host machine, you can find the address of your VM by running ``multipass info mattermost-tutorial-vm``.
   This would be the private IP address and usually would be listed first. Now, you can navigate to ``<your-multipass-vm-ip>:8080`` to access Mattermost.

Clean up the environment
~~~~~~~~~~~~~~~~~~~~~~~~~

Congratulations! You have successfully finished the Mattermost tutorial. You have now deployed the Mattermost charm, integrated it with a database, and accessed the Mattermost instance in your browser. You can now remove the
model environment that you've created using the following command:

.. code-block:: bash

   juju destroy-model mattermost-tutorial --destroy-storage

> If you used Multipass, to remove the Multipass instance you created for this tutorial, use the following command on your host machine.

Reviewer: ```suggestion
If you used Multipass, remove the instance you created for this tutorial using the following command on your host machine.
```
Nit
